OSC colour queries: terminate the reply with the same sequence used in
the query (ST, ESC \ or BEL), from upstream.

24-bit colour, OSC 110/111/112 colour reset.

Index: src/rxvt.h
--- src/rxvt.h.orig
+++ src/rxvt.h
@@ -41,7 +41,11 @@
 #else
 typedef uint16_t text_t; // saves lots of memory
 #endif
+#if USE_24_BIT_COLOR
+typedef uint64_t rend_t;
+#else
 typedef uint32_t rend_t;
+#endif
 typedef  int32_t tlen_t;  // was int16_t, but this results in smaller code and memory use
 typedef  int32_t tlen_t_; // specifically for use in the line_t structure
 
@@ -351,46 +355,55 @@
 #define RS_None                 0
 
 // GET_BGATTR depends on RS_fgShift > RS_bgShift
-#define RS_colorMask		((1UL << Color_Bits) - 1UL)
-#define RS_bgShift		0
-#define RS_fgShift		(RS_bgShift + Color_Bits)
-#define RS_bgMask               (RS_colorMask << RS_bgShift)
-#define RS_fgMask               (RS_colorMask << RS_fgShift)
+#define RS_colorMask		    ((rend_t) ((1UL << Color_Bits) - 1UL))
+#define RS_bgShift		        0
+#define RS_fgShift		        (RS_bgShift + Color_Bits)
+#define RS_bgMask               ((rend_t) (RS_colorMask << RS_bgShift))
+#define RS_fgMask               ((rend_t) (RS_colorMask << RS_fgShift))
 
 // must have space for rxvt_fontset::fontCount * 2 + 2 values
 #define RS_fontShift            (RS_fgShift + Color_Bits)
-#define RS_Careful		(1UL << RS_fontShift)	/* be careful when drawing these */
-#define RS_fontCount		rxvt_fontset::fontCount
-#define RS_fontMask             ((RS_fontCount << (RS_fontShift + 1)) | RS_Careful)   // includes RS_Careful
+#define RS_Careful		        ((rend_t) (1UL << RS_fontShift))	/* be careful when drawing these */
+#define RS_fontCount		    ((rend_t) rxvt_fontset::fontCount)
+#define RS_fontMask             ((rend_t) ((RS_fontCount << (RS_fontShift + 1)) | RS_Careful))   // includes RS_Careful
 
 // toggle this to force redraw, must be != RS_Careful and otherwise "pretty neutral"
-#define RS_redraw		(2UL << RS_fontShift)
+#define RS_redraw		        ((rend_t) (2UL << RS_fontShift))
+
+#if USE_24_BIT_COLOR
+# define RS_fontCountSize 4
+#elif USE_256_COLORS
+# define RS_fontCountSize 4
+#else
+# define RS_fontCountSize 8
+#endif
 
-#define RS_Sel                  (1UL << 22)
+#define RS_selShift             (RS_fontShift + RS_fontCountSize)
+#define RS_Sel                  ((rend_t) (1UL << RS_selShift))
 
 // 4 custom bits for extensions
 #define RS_customCount          16UL
-#define RS_customShift          23
-#define RS_customMask           ((RS_customCount - 1UL) << RS_customShift)
+#define RS_customShift          (RS_selShift + 1)
+#define RS_customMask           ((rend_t) ((RS_customCount - 1UL) << RS_customShift))
 
 // font styles
-#define RS_Bold                 (1UL << RS_styleShift)
-#define RS_Italic		(2UL << RS_styleShift)
+#define RS_Bold                 ((rend_t) (1UL << RS_styleShift))
+#define RS_Italic		        ((rend_t) (2UL << RS_styleShift))
 
-#define RS_styleCount		4
-#define RS_styleShift		27
-#define RS_styleMask		(RS_Bold | RS_Italic)
+#define RS_styleCount		    4
+#define RS_styleShift	        (RS_customShift + RS_styleCount)
+#define RS_styleMask		    ((rend_t) (RS_Bold | RS_Italic))
 
 // fake styles
-#define RS_Blink                (1UL << 29)
-#define RS_RVid                 (1UL << 30)    // reverse video
-#define RS_Uline                (1UL << 31)    // underline
+#define RS_Blink                ((rend_t) (1UL << (RS_styleShift + 2)))
+#define RS_RVid                 ((rend_t) (1UL << (RS_styleShift + 3)))    // reverse video
+#define RS_Uline                ((rend_t) (1UL << (RS_styleShift + 4)))    // underline
 
-#define RS_baseattrMask         (RS_Italic | RS_Bold | RS_Blink | RS_RVid | RS_Uline)
-#define RS_attrMask             (RS_baseattrMask | RS_fontMask)
+#define RS_baseattrMask         ((rend_t) (RS_Italic | RS_Bold | RS_Blink | RS_RVid | RS_Uline))
+#define RS_attrMask             ((rend_t) (RS_baseattrMask | RS_fontMask))
 
 #define DEFAULT_RSTYLE  (RS_None | (Color_fg    << RS_fgShift) | (Color_bg     << RS_bgShift))
-#define OVERLAY_RSTYLE  (RS_None | (Color_Black << RS_fgShift) | (Color_Yellow << RS_bgShift))
+#define OVERLAY_RSTYLE  (RS_None | (Color_White << RS_fgShift) | (Color_Blue << RS_bgShift))
 
 enum {
   C0_NUL = 0x00,
@@ -399,6 +412,12 @@
   C0_DLE, C0_DC1, C0_DC2, D0_DC3, C0_DC4, C0_NAK, C0_SYN, C0_ETB,
   C0_CAN, C0_EM , C0_SUB, C0_ESC, C0_IS4, C0_IS3, C0_IS2, C0_IS1,
 };
+
+struct string_term
+{
+  unsigned char v[3];
+};
+
 #define CHAR_ST                 0x9c    /* 0234 */
 
 /*
@@ -406,54 +425,58 @@
  * colour extensions by Christian W. Zuckschwerdt <zany@triq.net>
  */
 enum {
-  XTerm_name             =  0,
-  XTerm_iconName         =  1,
-  XTerm_title            =  2,
-  XTerm_property         =  3,      // change X property
-  XTerm_Color            =  4,      // change colors
-  XTerm_Color00          = 10,      // change fg color
-  XTerm_Color01          = 11,      // change bg color
-  XTerm_Color_cursor     = 12,      // change actual 'Cursor' color
-  XTerm_Color_pointer_fg = 13,      // change actual 'Pointer' fg color
-  XTerm_Color_pointer_bg = 14,      // change actual 'Pointer' bg color
-  XTerm_Color05          = 15,      // not implemented (tektronix fg)
-  XTerm_Color06          = 16,      // not implemented (tektronix bg)
-  XTerm_Color_HC         = 17,      // change actual 'Highlight' bg color
-  XTerm_Color_HTC        = 19,      // change actual 'Highlight' fg color
-  XTerm_logfile          = 46,      // not implemented
-  XTerm_font             = 50,
-
-  XTerm_konsole30        = 30,      // reserved for konsole
-  XTerm_konsole31        = 31,      // reserved for konsole
-  XTerm_emacs51          = 51,      // reserved for emacs shell
+  XTerm_name                =  0,
+  XTerm_iconName            =  1,
+  XTerm_title               =  2,
+  XTerm_property            =  3,      // change X property
+  XTerm_Color               =  4,      // change colors
+  XTerm_Color00             = 10,      // change fg color
+  XTerm_Color01             = 11,      // change bg color
+  XTerm_Color_cursor        = 12,      // change actual 'Cursor' color
+  XTerm_Color_pointer_fg    = 13,      // change actual 'Pointer' fg color
+  XTerm_Color_pointer_bg    = 14,      // change actual 'Pointer' bg color
+  XTerm_Color05             = 15,      // not implemented (tektronix fg)
+  XTerm_Color06             = 16,      // not implemented (tektronix bg)
+  XTerm_Color_HC            = 17,      // change actual 'Highlight' bg color
+  XTerm_Color_HTC           = 19,      // change actual 'Highlight' fg color
+  XTerm_logfile             = 46,      // not implemented
+  XTerm_font                = 50,
+
+  XTerm_konsole30           = 30,      // reserved for konsole
+  XTerm_konsole31           = 31,      // reserved for konsole
+  XTerm_emacs51             = 51,      // reserved for emacs shell
   /*
    * rxvt extensions of XTerm OSCs: ESC ] Ps;Pt (ST|BEL)
    */
 
   // deprecated
-  Rxvt_restoreFG         = 39,
-  Rxvt_restoreBG         = 49,
-
-  Rxvt_dumpscreen        = 55,      // dump scrollback and all of screen
-
-  URxvt_locale           = 701,     // change locale
-  URxvt_version          = 702,     // request version
-
-  URxvt_Color_IT         = 704,     // change actual 'Italic' colour
-  URxvt_Color_BD         = 706,     // change actual 'Bold' color
-  URxvt_Color_UL         = 707,     // change actual 'Underline' color
-  URxvt_Color_border     = 708,
+  Rxvt_restoreFG            = 39,
+  Rxvt_restoreBG            = 49,
 
-  URxvt_font             = 710,
-  URxvt_boldFont         = 711,
-  URxvt_italicFont       = 712,
-  URxvt_boldItalicFont   = 713,
+  Rxvt_dumpscreen           = 55,      // dump scrollback and all of screen
+                                       //
+  XTerm_RestoreColor00      = 110,     // restore fg color
+  XTerm_RestoreColor01      = 111,     // restore bg color
+  XTerm_RestoreColor_cursor = 112,     // restore cursor color
+
+  URxvt_locale              = 701,     // change locale
+  URxvt_version             = 702,     // request version
+
+  URxvt_Color_IT            = 704,     // change actual 'Italic' colour
+  URxvt_Color_BD            = 706,     // change actual 'Bold' color
+  URxvt_Color_UL            = 707,     // change actual 'Underline' color
+  URxvt_Color_border        = 708,
+
+  URxvt_font                = 710,
+  URxvt_boldFont            = 711,
+  URxvt_italicFont          = 712,
+  URxvt_boldItalicFont      = 713,
 
-  URxvt_view_up          = 720,
-  URxvt_view_down        = 721,
+  URxvt_view_up             = 720,
+  URxvt_view_down           = 721,
 
-  URxvt_cellinfo         = 776,     // returns font cell width, height etc.
-  URxvt_perl             = 777,     // for use by perl extensions, starts with "extension-name;"
+  URxvt_cellinfo            = 776,     // returns font cell width, height etc.
+  URxvt_perl                = 777,     // for use by perl extensions, starts with "extension-name;"
 };
 
 /* Words starting with `Color_' are colours.  Others are counts */
@@ -538,7 +561,9 @@
 #endif
 };
 
-#if USE_256_COLORS
+#if USE_24_BIT_COLOR
+# define Color_Bits      25
+#elif USE_256_COLORS
 # define Color_Bits      9 // 0 .. maxTermCOLOR24
 #else
 # define Color_Bits      8 // 0 .. maxTermCOLOR24
@@ -674,21 +699,21 @@
 #define ROW(n) ROW_of (this, n)
 
 /* how to build & extract colors and attributes */
-#define GET_BASEFG(x)           (((x) & RS_fgMask) >> RS_fgShift)
-#define GET_BASEBG(x)           (((x) & RS_bgMask) >> RS_bgShift)
+#define GET_BASEFG(x)           ((((rend_t) (x)) & RS_fgMask) >> RS_fgShift)
+#define GET_BASEBG(x)           ((((rend_t) (x)) & RS_bgMask) >> RS_bgShift)
 
-#define GET_FONT(x)             (((x) & RS_fontMask) >> RS_fontShift)
-#define SET_FONT(x,fid)         (((x) & ~RS_fontMask) | ((fid) << RS_fontShift))
+#define GET_FONT(x)             ((((rend_t) (x)) & RS_fontMask) >> RS_fontShift)
+#define SET_FONT(x,fid)         ((((rend_t) (x)) & ~((rend_t) RS_fontMask)) | (((rend_t) (fid)) << RS_fontShift))
 
-#define GET_STYLE(x)		(((x) & RS_styleMask) >> RS_styleShift)
-#define SET_STYLE(x,style)	(((x) & ~RS_styleMask) | ((style) << RS_styleShift))
+#define GET_STYLE(x)		    ((((rend_t) (x)) & RS_styleMask) >> RS_styleShift)
+#define SET_STYLE(x,style)	    ((((rend_t) (x)) & ~((rend_t) RS_styleMask)) | (((rend_t) (style)) << RS_styleShift))
 
-#define GET_ATTR(x)             (((x) & RS_attrMask))
-#define SET_FGCOLOR(x,fg)       (((x) & ~RS_fgMask)   | ((fg) << RS_fgShift))
-#define SET_BGCOLOR(x,bg)       (((x) & ~RS_bgMask)   | ((bg) << RS_bgShift))
-#define SET_ATTR(x,a)           (((x) & ~RS_attrMask) | (a))
+#define GET_ATTR(x)             ((((rend_t) (x)) & RS_attrMask))
+#define SET_FGCOLOR(x,fg)       ((((rend_t) (x)) & ~((rend_t) RS_fgMask))   | (((rend_t) (fg)) << RS_fgShift))
+#define SET_BGCOLOR(x,bg)       ((((rend_t) (x)) & ~((rend_t) RS_bgMask))   | (((rend_t) (bg)) << RS_bgShift))
+#define SET_ATTR(x,a)           ((((rend_t) (x)) & ~((rend_t) RS_attrMask)) | ((rend_t) (a)))
 
-#define RS_SAME(a,b)		(!(((a) ^ (b)) & ~RS_Careful))
+#define RS_SAME(a,b)		    (!((((rend_t) (a)) ^ ((rend_t) (b))) & ~((rend_t) RS_Careful)))
 
 #define PIXCOLOR_NAME(idx)      rs[Rs_color + (idx)]
 #define ISSET_PIXCOLOR(idx)     (!!rs[Rs_color + (idx)])
@@ -1222,6 +1247,8 @@
 #endif
 
   const char     *rs[NUM_RESOURCES];
+  const char     *def_colors[NRS_COLORS];
+
   /* command input buffering */
   char           *cmdbuf_ptr, *cmdbuf_endp;
   char            cmdbuf_base[CBUFSIZ];
@@ -1366,11 +1393,11 @@
   void process_escape_seq ();
   void process_csi_seq ();
   void process_window_ops (const int *args, unsigned int nargs);
-  char *get_to_st (unicode_t &ends_how);
+  char *get_to_st (string_term &st);
   void process_dcs_seq ();
   void process_osc_seq ();
-  void process_color_seq (int report, int color, const char *str, char resp);
-  void process_xterm_seq (int op, char *str, char resp);
+  void process_color_seq (int report, int color, const char *str, string_term &st);
+  void process_xterm_seq (int op, char *str, string_term &st);
   unsigned int map_rgb24_color (unsigned int r, unsigned int g, unsigned int b, unsigned int a);
   int privcases (int mode, unsigned long bit);
   void process_terminal_mode (int mode, int priv, unsigned int nargs, const int *arg);
@@ -1411,6 +1438,7 @@
   void set_window_color (int idx, const char *color);
   char *get_colorfgbg ();
   bool set_color (rxvt_color &color, const char *name);
+  const char *get_def_color(int idx);
   void alias_color (int dst, int src);
   void set_widthheight (unsigned int newwidth, unsigned int newheight);
   void get_window_origin (int &x, int &y);
@@ -1488,7 +1516,12 @@
   void scr_swap_screen () noexcept;
   void scr_change_screen (int scrn);
   void scr_color (unsigned int color, int fgbg) noexcept;
-  void scr_rendition (int set, int style) noexcept;
+#if USE_24_BIT_COLOR
+  void scr_color_24 (unsigned int color, int fgbg) noexcept;
+  void scr_color_rgb (unsigned int r, unsigned int g, unsigned int b, int fgbg) noexcept;
+#endif
+  rxvt_color &lookup_color (unsigned int color, rxvt_color *table) noexcept;
+  void scr_rendition (int set, rend_t style) noexcept;
   void scr_add_lines (const wchar_t *str, int len, int minlines = 0) noexcept;
   void scr_backspace () noexcept;
   void scr_tab (int count, bool ht = false) noexcept;
